count_blocklist.go

```diff
@@ -0,0 +1,66 @@
+package p2p
+
+import (
+	"time"
+
+	"github.com/iotexproject/go-pkgs/cache"
+	"github.com/libp2p/go-libp2p-core"
+)
+
+// CountBlocklist is a count-based blacklist implementation. Entry is being added
+// to an LRU cache, once the count reaches a threshold the entry is blocked for a
+// certain timeout. The entry is then unblocked once the timeout expires
+type CountBlocklist struct {
+	count   int           // number of times an entry is added before being blocked
+	ttl     time.Duration // timeout an entry will be blocked
+	counter *cache.ThreadSafeLruCache
+	timeout *cache.ThreadSafeLruCache
+}
+
+// NewCountBlocklist creates a new CountBlocklist
+func NewCountBlocklist(cap, count int, ttl time.Duration) *CountBlocklist {
+	return &CountBlocklist{
+		count:   count,
+		ttl:     ttl,
+		counter: cache.NewThreadSafeLruCache(cap),
+		timeout: cache.NewThreadSafeLruCache(cap),
+	}
+}
+
+// Blocked returns true if the name is blocked
+func (b *CountBlocklist) Blocked(name core.PeerID, t time.Time) bool {
```

Comment:
why we need "t" as a parameter?

Reply:
to determine if timeout has expired, say the name is added at time `t0`

if `t < t0 + 10min`, it returns true, otherwise it returns false (since 10 min has passed)

Reply:
t which is always "now" could be local parameter.

Reply:
yes, for now, t is always "now"
but here I want to pass in time explicitly and remove that assumption/dependency, so in theory, i can do
```
Add(name, 0)
Blocked(name, 0+10min)
```
vs.
```
Add(name)
Blocked(name)
and use time.Now() inside
```